Add contract tests for OpenRosaHttpInterface

seadowg opened the pull request:
Closes #2869

<!-- 
Thank you for contributing to ODK Collect!

Before sending this PR, please read
https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md
-->

#### What has been done to verify that this works as intended?

#### Why is this the best possible solution? Were any other approaches considered?

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

#### Do we need any specific form for testing your changes? If so, please attach one.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

#### Before submitting this PR, please make sure you have:
- [ ] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [ ] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [ ] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

seadowg committed:
Add basic tests for OpenRosaHttpInterface.executeGetRequest

seadowg committed:
Fix non final field in test

seadowg committed:
Add header and content type tests for OpenRosaInterface.executeGetRequest

seadowg committed:
Add tests for gzip non 200 responses

seadowg commented:
Going to keep a log of discrepancies/weirdness I run into. Starting now:

❓ ~~Old implementation functions fine with non standard content types where as new implementation can't handle them (they simply become `null`). This is probably ok - I've ignored the test for now so we can discuss.~~

❓ ~~New implementation sends `gzip,deflate` instead of just `gzip` as its `Accepts-Encoding` header. Again this is most likely not a problem -  I've ignored the test for now so we can discuss.~~

🤷‍♂️ Original implementation has path that throws an `Exception` if the response body is null. It seems this will never actually get executed however as `204` and `304` responses are returned as an `HttpGetResult` via an earlier status code check. I've made sure there is a test covering the actual behavior.

seadowg committed:
Add tests for Basic auth behaviour for get requests

seadowg commented on collect_app/src/test/java/org/odk/collect/android/http/OpenRosaHttpInterfaceTest.java:
```diff
@@ -86,6 +90,36 @@ public void executeGetRequest_sendsAcceptsGzipHeader() throws Exception {
     }
 
+    @Test
+    @Ignore("OkHttpConnection not fooled into https behaviour by port 8443 - scheme needs to be https")
```
This is a weird one: Basically the old Apache based implementation would send basic authentication credentials after a challenge if the connection port was either 443 or 8443 (standard https ports) where as the new OkHttp implementation. This means that technically you could do basic auth over plain ol' HTTP running on an HTTPS standard port.

I'm guessing from reading the [Open Rosa Authentication docs](https://docs.opendatakit.org/openrosa-authentication/) that the intention is for basic auth to only happen over HTTPS and that the no server/client setups would be relying on using basic auth over plain HTTP (hackingly) setup on port 8443? Does that sound right @ggalmazor @yanokwa?

xsteelej replied:
Hi @seadowg, just to clarify, are you saying that the OkHttp implementation will not send basic credentials if using "http" even if the connection port is 443 or 8443?

To me, this sounds like correct behaviour - see the "http" check in OkHttpConnection.setCredentialsIfNeeded.

Thanks

lognaturel replied:
Agreed that basic auth credentials should only be sent if the scheme is HTTPS regardless of connection port. 👍

seadowg replied:
Yeah that's right @xsteelej. I agree with yourself and @lognaturel here!

I don't want to start changing the behavior of the original implementation to be smarter and actually check the protocol. I've been trying to work on having a slightly different "fake" https server for each each implementation's test. 

The original implementation is super easy because I can trick it with port `8443` but it's harder to get https tests working for the new implementation as it's more correct and actually needs a localhost https server running. Going to reach out to some OkHttp/MockWebServer folks for some help!

seadowg committed:
Make content encoding consistent for Open Rosa implementations

seadowg commented:
Oh yeah I totally agree it seems correct!

I think it's a weird behaviour of the original implementation - that it
sends basic auth credentials on https ports even when the protocol is http.
I just wanted to double check it wasn't intended.

On Wed, 13 Mar 2019, 22:49 xsteelej, <notifications@github.com> wrote:

> *@xsteelej* commented on this pull request.
> ------------------------------
>
> In
> collect_app/src/test/java/org/odk/collect/android/http/OpenRosaHttpInterfaceTest.java
> <https://github.com/opendatakit/collect/pull/2917#discussion_r265362056>:
>
> > @@ -86,6 +90,36 @@ public void executeGetRequest_sendsAcceptsGzipHeader() throws Exception {
>          assertThat(request.getHeader("Accept-Encoding"), equalTo("gzip"));
>      }
>
> +    @Test
> +    @Ignore("OkHttpConnection not fooled into https behaviour by port 8443 - scheme needs to be https")
>
> Hi @seadowg <https://github.com/seadowg>, just to clarify, are you saying
> that the OkHttp implementation will not send basic credentials if using
> "http" even if the connection port is 443 or 8443?
>
> To me, this sounds like correct behaviour - see the "http" check in
> OkHttpConnection.setCredentialsIfNeeded.
>
> Thanks
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opendatakit/collect/pull/2917#discussion_r265362056>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAh8-N7Z_P0O44zLc-fljjjGYA4yuODSks5vWYBmgaJpZM4bR8ij>
> .
>

seadowg committed:
Allow OpenRosaHttpInterfaceTest to test https behaviour

Sets up MockWebServer to use localhost https config. Also needs
to inject OkHttpClient Builder into OkHttpConnection so it can
be configured to use localhost https config.

seadowg committed:
Move https tests for OpenRosaInterface into client specific tests

It feels like the behaviour (although the intention is the same)
is different enough to warrant the tests being implementation
specific

seadowg commented:
Finding from today's debugging:

❓ There is a specific piece of code in the `executeGetRequest` method of `HttpClientConnection` that deals with gzipped responses:

```java
if (contentEncoding != null && contentEncoding.getValue().equalsIgnoreCase(GZIP_CONTENT_ENCODING)) {
    downloadStream = new GZIPInputStream(downloadStream);
}
```

It appears that this will never get called as far as I can see however as Apache Client (like OkHttp) is dealing with the gzip decompression for us automatically. Frustratingly it also seems the Apache Client takes it upon itself to remove the `Content-Encoding` header which has made today longer than it needed to be 😒.

I'm working on a generalized test for gzip functionality as this is important but as I would have expected this is something we get for free from both http frameworks.

seadowg committed:
Add comments to call out dead code

seadowg committed:
Remove unused imports

seadowg committed:
Remove bogus content type test for Open Rosa interface

Assume we don't care about behaviour if server returns unrecognized
content types

seadowg committed:
Stop explicitly setting gzip encoding for OkHttpConnection requests

If we set the head explicitly we lost OkHttp's automagic gzip
header sending and decompression.

seadowg committed:
Remove comment moved from test to implementation

seadowg committed:
Seperate open rosa interface tests by method

seadowg committed:
Add basic tests for Open Rosa HEAD request

seadowg committed:
Simplify Open Rosa test names

seadowg committed:
Remove unused executePostRequest method from OpenRosaHttpInterface

This seems to have been added so the interface is just concerned with
HTTP verbs rather than different actions (such as uploading a form).
Removing the method for the moment to keep the interface simpler until
we've switched to OkHttp.

seadowg committed:
Add cookie clearing test to Open Rosa interface tests

seadowg committed:
Add test to verify Open Rosa head request exception throwing

seadowg committed:
Add tests for OkHttpConnection head request

seadowg committed:
Add basic tests for Open Rosa post requests

seadowg committed:
Add error case test for Open Rosa post requests

yanokwa commented:
@lognaturel Are you sure you wanted this closed?

lognaturel commented:
No, definitely not!

seadowg committed:
Add test skeletons for post request behaviour

seadowg committed:
Add fail() to test bodies so they are not empty

seadowg committed:
Implement submission file test for Open Rosa POST

seadowg committed:
Add tests for Open Rosa POST request statuses

seadowg committed:
Add tests for attachment sending with Open Rosa POST

seadowg committed:
Add HTTPS test for Open Rosa POST

seadowg committed:
Fix OkHttpConnection nullpointerexception when parsing content type

seadowg committed:
Extract commom content type mapping from Open Rosa clients

seadowg committed:
Inject fake content mapper in Open Rosa POST tests

seadowg committed:
Fix broken HttpClientConnectiont contruction in tests

seadowg committed:
Fix style problems

seadowg committed:
Add test for split requests during Open Rosa POST

seadowg committed:
Add split request failure case tests for Open Rosa POST

seadowg commented:
@lognaturel I think we're at a good stage to start thinking about merging this. The tests don't cover every last detail but it looks like we have almost all the behavior specified. I've only made some very small changes to `HttpClientConnection` (mostly as comments) and just a few changes to OkHttpClientConnection where we didn't have parity.

One last thing would be digest authentication but I feel like we can most likely trust any underlying HTTP framework and I'm assuming this would be tested by QA/Beta. I'm not sure I'll have time to go down that route.

Let me know you're thoughts on what steps we might want to take before merging. A core question would be as to whether we just merge the tests in first then make the call to switch clients at a later date or do that as part of this PR now we have the coverage.

EDIT: Just to clarify there is still a lot of work that could be continued with this particular train of thought. However, it doesn't seem a lot of people are touching this code and improving it wouldn't (as far as I can see) have benefits to the end user so I'm not sure it should be a priority 🤷‍♂️

lognaturel commented on collect_app/src/test/java/org/odk/collect/android/http/OpenRosaHeadRequestTest.java:
```diff
@@ -107,4 +111,15 @@ public void withCredentials_whenHttp_doesNotRetryWithCredentials() throws Except
+            subject.executeHeadRequest(new URI("http://localhost:8443"), null);
+            fail();
+        } catch (Exception e) {
+            assertThat(e, isA(Exception.class));
```
It's impossible to get here without the assertion being true, right?

seadowg replied:
Ah yeah I think I initially had a more granular (with respect to the exception type) assertion here and then updated it. You're totally right that now it doesn't need to be there. Derp. I'll fix and then resolve this.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/http/ContentTypeMapping.java:
```diff
@@ -0,0 +1,42 @@
+import org.opendatakit.httpclientandroidlib.entity.ContentType;
+
+public enum ContentTypeMapping {
+    XML("xml",  ContentType.TEXT_XML),
```
Not related to your changes but now that this is pulled out I'm noticing that the list should probably be reduced to the extensions that are not in Android's `MimeTypeMap` -- `amr`, `oga`, etc. Or I'd even lean towards saying those are obscure/obsolete enough that we could do away with them altogether and let them go through as a generic octet stream.

Aside from that, I'd prefer replacing the whole `enum` with a conditional/switch in your method that goes from filename to mime type. This whole logic feels unnecessarily complicated.

(Maybe if we pull the code apart method by method we'll get to something simpler. Hah.)

seadowg replied:
Yeah I had to sit on my hands to stop myself from playing around with it. What's nice now is the content mapping logic can be unit tested which will allow us to safely make changes to it.

lognaturel replied:
> Yeah I had to sit on my hands to stop myself from playing around with it.

Seriously.

How about filing an issue that we can tag as `good first issue` that describes in some detail what test to write and how to simplify this? It's always nice to have some of these non-urgent things for new folks to try their hands at.

lognaturel commented on collect_app/src/test/java/org/odk/collect/android/http/OpenRosaPostRequestTest.java:
```diff
@@ -0,0 +1,323 @@
+        return Arrays.stream(stringParts).map(part -> part.split("\r\n")).collect(Collectors.toList());
+    }
+
+    private class XmlOrBlahContentTypeMapper implements OpenRosaHttpInterface.FileToContentTypeMapper {
```
You need this because of `MimeTypeMap` being in Android, right?

seadowg replied:
Yeah that was the initial motivation for extracting it. Another nicety about having a fake here is it lets us not worry about any of the complexities of the actual content mapper when we're writing the tests.

lognaturel reviewed (commented):
Thanks, @seadowg. Looks like really thorough and thoughtful work to me. I have a couple of not super important questions in line. Agreed not having specific digest auth tests is ok because of your points about QA and because the OKHttp digest lib seems reasonably well-tested.

Can you please make it not a draft? Let's merge the tests now and you can follow up with another PR to flip the switch.

Some things I thought about going through this:
- We should make sure to verify the redirect behavior. I don't think there needs to be a test at this layer because the (probably should be reconsidered) logic is in `InstanceServerUploader`. At this level redirects are just enabled.
- I remember some odd credentials caching when the OkHttp code was being worked through. Again, I don't think there's a test to write at this level but we'll probably want to double check that things work as expected in cases like doing some submissions then switching credentials and/or switching server. Probably worth doing for both basic and digest?

Those can be checked by QA when the switch get flipped.

seadowg committed:
Remove useless assertion

seadowg committed:
Remove unused exception

seadowg commented:
@lognaturel it's a real PR now. The credential caching sounds interesting. If we find anything with QA it'd be interesting to think about whether we need more tests at this level.

I'll do a rebase on this so it's all good to merge but hang off on the OKHttp switch PR until we've got it in.
